File: C++/54.螺旋矩阵.cpp

// 54. 螺旋矩阵
//
// 给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。

#include <vector>

class Solution {
public:
    std::vector<int> spiralOrder(std::vector<std::vector<int>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0) {
            return {};
        }

        // 总结果
        std::vector<int> result;
        result.reserve(matrix.size() * matrix[0].size());

        // 要打印的圈数
        int order_num = std::min(matrix.size() + 1, matrix[0].size() + 1) / 2;

        // 打印每一圈
        for (int i = 0; i < order_num; i++) {
            printOrder(matrix, i, result);
        }

        return result;
    }

    // 打印一圈
    void printOrder(std::vector<std::vector<int>>& matrix, int start, std::vector<int>& result) {
        int endX = matrix[0].size() - start - 1;  // 该圈最大横坐标
        int endY = matrix.size() - start - 1;     // 该圈最大纵坐标

        // 打印上面一行
        for (int i = start; i <= endX; i++) {
            result.emplace_back(matrix[start][i]);
        }

        // 打印右面一列
        for (int j = start + 1; j <= endY; j++) {
            result.emplace_back(matrix[j][endX]);
        }

        // 打印下面一行
        if (endY > start) {
            for (int i = endX - 1; i >= start; i--) {
                result.emplace_back(matrix[endY][i]);
            }
        }

        // 打印左面一列
        if (endX > start) {
            for (int j = endY - 1; j > start; j--) {
                result.emplace_back(matrix[j][start]);
            }
        }
    }
};

File: leetcode/01数组/54. 螺旋矩阵.go
package main

// 54. 螺旋矩阵

// 给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。

// spiralOrder .
func spiralOrder(matrix [][]int) []int {
	result := make([]int, 0, len(matrix)*len(matrix[0])) // 总结果
	orderNum := min(len(matrix)+1, len(matrix[0])+1) / 2 // 要打印的圈数

	// 打印每一圈
	for i := 0; i < orderNum; i++ {
		printOrder(matrix, i, &result)
	}

	return result
}

// printOrder 打印一圈
func printOrder(matrix [][]int, start int, result *[]int) {
	endX := len(matrix[0]) - start - 1 // 该圈最大横坐标
	endY := len(matrix) - start - 1    // 该圈最大纵坐标

	// 打印上面行
	for j := start; j <= endX; j++ {
		*result = append(*result, matrix[start][j])
	}

	// 打印右面列
	for i := start + 1; i <= endY; i++ {
		*result = append(*result, matrix[i][endX])
	}

	// 打印下面行 -- 需要判断下是否只有一行，防止只有一行时被打印多次
	if endY > start {
		for j := endX - 1; j >= start; j-- {
			*result = append(*result, matrix[endY][j])
		}
	}

	// 打印左面列 -- 需要判断下是否只有一列，防止只有一列时被打印多次
	if endX > start {
		for i := endY - 1; i > start; i-- {
			*result = append(*result, matrix[i][start])
		}
	}
}
